interfaces/policy/policy.go

```diff
@@ -116,22 +116,23 @@ func (ic *InstallCandidate) Check() error {
 
 // ConnectCandidate represents a candidate connection.
 type ConnectCandidate struct {
-	// TODO: later we need to carry dynamic attributes once we have those
 	Plug                *snap.PlugInfo
 	PlugSnapDeclaration *asserts.SnapDeclaration
+	PlugAttrs           map[string]interface{}
 
 	Slot                *snap.SlotInfo
 	SlotSnapDeclaration *asserts.SnapDeclaration
+	SlotAttrs           map[string]interface{}
 
 	BaseDeclaration *asserts.BaseDeclaration
 }
 
 func (connc *ConnectCandidate) plugAttrs() map[string]interface{} {
-	return connc.Plug.Attrs
+	return connc.PlugAttrs
```

Comment:
ConnectCandidate is also meant to write easily tests about the correctness of decls, so I think I would prefer this to have a fallback:

    if connc.PlugAttrs != nil {
        return connc.PlugAttrs
    }
   return connc.Plug.Attrs

and then have in policy just a couple of tests that test the new paths